### ISSUE TITLE
Updating coredistools

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/jitutils#414]
What should be included in the next update of the coredistools library (used for x86/x64 GC stress, R2RDump disassembly, superpmi asm diffs, RyuJIT "late disassembler", ILC (?), and possibly more)?

Some proposals:
1. Update LLVM from 17.0.6 to 19.1.0 (or whatever is latest when we finally update): https://github.com/llvm/llvm-project/releases.
2. Add LoongArch and RISC-V disassemblers to the standard set built on win-x64, win-arm64, etc. (Allows for superpmi asm diffs using cross compilers, for example.)
3. Does the latest LLVM contain support for the latest AVX-512? AVX10? [Intel APX](https://www.intel.com/content/www/us/en/developer/articles/technical/advanced-performance-extensions-apx.html)? Arm SVE/SVE2/SSVE/SME? Other new ISA extensions? (If not, is new LLVM ISA support for desired ISAs expected soon?)

Cordistools.cpp currently tweaks the "features string" when creating the disassembler, to create a better set of ISAs for disassembly:
```
  if (TheTargetArch == Target_Arm64) {
    // Enable all features for disassembly. Setting a specific advanced CPU enables all the architecture
    // features for that CPU (e.g., `Mcpu = "neoverse-n2"`), but we want to use the "meta" feature
    // string "+all" to just enable all features, even those not implemented in any current CPU.
    FeaturesStr = "+all";
  } else if (TheTargetArch == Target_RiscV64) {
    FeaturesStr = "+m,+a,+f,+d,+c,+zicsr,+zifencei";  // RV64GC
  }
```
Notably, when this was done, "+all" for arm64 was not in "great" shape.

Is this still the correct set, for LLVM 19.1.0 (and beyond)?

Is there any pressing need to either update coredistools, or wait to update it?

Comments?

@dotnet/jit-contrib @dotnet/samsung @shushanhf @dotnet/crossgen-contrib @khushal1996 @DeepakRajendrakumaran @anthonycanino

**Comment by shushanhf** [comment]
> 2. Add LoongArch and RISC-V disassemblers to the standard set built on win-x64, win-arm64, etc. (Allows for superpmi asm diffs using cross compilers, for example.)

@BruceForstall 
Thanks very much for adding LoongArch.
For LoongArch, if there are some work to do, you can tell me to do that.

**Comment by BruceForstall** [comment]
Also @dotnet/arm64-contrib

**Comment by tomeksowi** [comment]
> Is there any pressing need to either update coredistools, or wait to update it?

Not really from RISC-V. We will be adding RVA22 profile extensions somewhere in 2025, but RV64GC good for now.

**Comment by tomeksowi** [comment]
> EDIT: I'll check if "+all" works now for RISC-V

Nope, it's AArch64 only: `'+all' is not a recognized feature for this target (ignoring feature)`

But while I'm here, I'll add more extensions for RISC-V to future-proof.

**Comment by anthonycanino** [comment]
Hi Bruce, I will check internally and get back to you on ISA support in LLVM latest. I believe by just using latest we should be fine, but will confirm whether we have any pending work.

**Comment by anthonycanino** [comment]
@BruceForstall latest LLVM disassembler should have full support for APX and AVX10.2. I think we are good.

**Comment by BruceForstall** [comment]
I've been trying to update the coredistools build, specifically updating to using the latest Azure Linux 3 .NET 10 containers (used by the dotnet/runtime build), as well as the latest LLVM 20.1.0, with this PR:

https://github.com/dotnet/jitutils/pull/412

The builds are failing. I probably can't spend much time on this, but if there is a Linux LLVM build expert willing to take a look at that PR, and even submit proposed changes to it, I'd love to get that help.

**Comment by a74nh** [comment]
> I've been trying to update the coredistools build, specifically updating to using the latest Azure Linux 3 .NET 10 containers (used by the dotnet/runtime build), as well as the latest LLVM 20.1.0, with this PR:
> 
> [#412](https://github.com/dotnet/jitutils/pull/412)
> 
> The builds are failing. I probably can't spend much time on this, but if there is a Linux LLVM build expert willing to take a look at that PR, and even submit proposed changes to it, I'd love to get that help.

Quick glance:

```
2025-03-18T18:19:42.5928417Z     cc: fatal error: cannot execute 'as': execvp: No such file or directory
```

it can't find the assembler. I suspect your either need to install an additional package or your version of `as` is something like `as-20.1` and a set-defaults script needs running somewhere to create a symbolic link.

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib @dotnet/arm64-contrib @dotnet/samsung @shushanhf @DeepakRajendrakumaran @anthonycanino

We now have a good CI build of the latest coredistools. Thus, I am planning to update the version we use in dotnet/runtime soon. Are there any changes that anyone want to make to coredistools before it gets updated?

If you want to try it out, download the .nupkg from:

https://dev.azure.com/dnceng-public/public/_build/results?buildId=1006193&view=artifacts&pathAsName=false&type=publishedArtifacts

rename it to .zip, uncompress it, and replace the coredistools.dll or libcordistools.so/dylib in Tests/Core_Root, and use it. 

For RISC-V, if you sync to the latest dotnet/runtime after https://github.com/dotnet/runtime/pull/114291, you can use superpmi's "asmdiffs" functionality even using cross-compilation (it doesn't work for everything). E.g.,
```
python3 src/coreclr/scripts/superpmi.py asmdiffs -target_arch riscv64 -f bench
```
If you have any RISC-V natively collected MCH files, you can of course use them.

**Comment by am11** [comment]
@tomeksowi, RV{A,B}23 are also ratified https://lf-riscv.atlassian.net/wiki/spaces/HOME/pages/16154769/RISC-V+Technical+Specifications#Profiles, shall we add them as well? Particularly zacas is something NativeAOT could use to skip depending on libatomic for CAS-128 (https://github.com/dotnet/runtime/blob/a2e1d21bb4faf914363968b812c990329ba92d8e/src/coreclr/nativeaot/Runtime/unix/PalRedhawkInline.h#L106).

**Comment by tomeksowi** [comment]
> [@tomeksowi](https://github.com/tomeksowi), RV{A,B}23 are also ratified https://lf-riscv.atlassian.net/wiki/spaces/HOME/pages/16154769/RISC-V+Technical+Specifications#Profiles, shall we add them as well?

I guess more extensions in disassembly wouldn't hurt, #419 

RVB seems to have less extensions than RVA so I didn't include explicitly.

**Comment by DeepakRajendrakumaran** [comment]
I'm able to verify that this `coredistools` works for APX. The ran the following 2 tests
1. ran superpmi asmdiffs with my local build. With current `coredistools`, this throws the following error : `ERROR: Decode Failure`. When I copy the `coredistools` from the PR, the error goes away
2. Ran emitter unit tests for APX with `REX2` and `eEVEX` encoding and relevant environment variables. Able to decode instructions

**Comment by BruceForstall** [comment]
Thank you @DeepakRajendrakumaran !

> Is there any pressing need to either update coredistools, or wait to update it? - It's a good to have for me since without this, superpmi asmdiff tests with APX enabled will have the decode error

I'm in the process of updating it now.

**Comment by BruceForstall** [comment]
Coredistools has been updated! The version number is now 1.5. The updated nuget package has been uploaded and the PR https://github.com/dotnet/runtime/pull/114507 updates dotnet/runtime to use it.
